TombEngine/Objects/TR3/Entity/tr3_wasp.cpp

```diff
@@ -0,0 +1,286 @@
+		WASP_ANIM_IDLE,
+		WASP_ANIM_IDLE_TO_FLY_IDLE,
+		WASP_ANIM_ATTACK,
+		WASP_ANIM_FALL,
+		WASP_ANIM_DEATH,
+		WASP_ANIM_FLY_FORWARD
+	};
+
+	static void SpawnWaspParticle(short itemNumber)
+	{
+		auto* particle = GetFreeParticle();
+
+		particle->on = true;
+		particle->sG = Random::GenerateInt(32, 96);
+		particle->sB = particle->sG >> 1;
+		particle->sR = particle->sG >> 2;
+
+		particle->dG = Random::GenerateInt(224, 256);
+		particle->dB = particle->dG >> 1;
+		particle->dR = particle->dG >> 2;
+
+		particle->colFadeSpeed = 4;
+		particle->fadeToBlack = 2;
+		particle->sLife = particle->life = 8;
+
+		particle->blendMode = BLEND_MODES::BLENDMODE_ADDITIVE;
+
+		particle->extras = 0;
+		particle->dynamic = -1;
+
+		particle->x = Random::GenerateInt(-8, 8);
+		particle->y = Random::GenerateInt(-8, 8);
+		particle->z = Random::GenerateInt(-64, 64);
+
+		particle->xVel = Random::GenerateInt(-32, 32);
+		particle->yVel = Random::GenerateInt(-32, 32);
+		particle->zVel = Random::GenerateInt(-32, 32);
+		particle->friction = 2 | (2 << 4);
+
+		particle->flags = SP_SCALE | SP_ITEM | SP_NODEATTACH | SP_DEF;
+		particle->gravity = particle->maxYvel = 0;
+
+		particle->fxObj = itemNumber;
+		particle->nodeNumber = ParticleNodeOffsetIDs::NodeWasp;
+
+		particle->spriteIndex = Objects[ID_DEFAULT_SPRITES].meshIndex;
+		particle->scalar = 3;
+
+		int size = Random::GenerateInt(4, 8);
+
+		particle->size =
+		particle->sSize = size;
+		particle->dSize = size >> 1;
+	}
+
+	static void DoWaspEffects(short itemNumber, ItemInfo& item)
+	{
+		constexpr auto PARTICLE_EFFECT_COUNT = 2;
+
+		// Spawn light.
+		auto pos = GetJointPosition(&item, WaspVenomSackBite.meshNum, WaspVenomSackBite.Position);
+		TriggerDynamicLight(
+			pos.x, pos.y, pos.z, 10, 
+			WaspVenomSackLightColor.x * 255,
+			WaspVenomSackLightColor.y * 255,
+			WaspVenomSackLightColor.z * 255);
+
+		// Spawn wasp effect twice.
+		for (int i = 0; i < PARTICLE_EFFECT_COUNT; i++)
+			SpawnWaspParticle(itemNumber);
+	}
+
+	void InitialiseWaspMutant(short itemNumber)
+	{
+		auto& item = g_Level.Items[itemNumber];
+		InitialiseCreature(itemNumber);
+		SetAnimation(&item, WASP_STATE_IDLE); // Start as flying idle.
+	}
+
+	// NOTE: AI_MODIFY not allow the wasp to land,
+	// and if he start at the land state (which is by default the case),
+	// he will be forced to fly !
+	void WaspMutantControl(short itemNumber)
+	{
+		if (!CreatureActive(itemNumber))
+			return;
+
+		auto& item = g_Level.Items[itemNumber];
+		auto& creature = *GetCreatureInfo(&item);
+
+		short headingAngle = 0;
+
+		if (item.HitPoints <= 0)
+		{
+			switch (item.Animation.ActiveState)
+			{
+			case WASP_STATE_FALL:
+				if (item.Pose.Position.y >= item.Floor)
+				{
+					item.Pose.Position.y = item.Floor;
+					item.Animation.IsAirborne = false;
+					item.Animation.Velocity.y = 0.0f;
+					item.Animation.TargetState = WASP_STATE_DEATH;
+				}
+
+				break;
+
+			case WASP_STATE_DEATH:
+				item.Pose.Position.y = item.Floor;
```

Comment:
I think this should be <= for safety.

Reply:
if it's here, you can't :x

Reply:
nevermind, it not answer the question, it's in line 146, it check for the floor height there and set the state WASP_STATE_DEATH, so no need :)